src/backend/executor/execUtils.c

```diff
@@ -2098,6 +2099,24 @@ void mppExecutorFinishup(QueryDesc *queryDesc)
 		ExecSquelchNode(queryDesc->planstate);
 	}
 
+	/*
+	 * If the reading part of SharedScan is in slice 0, then we will not notify
+	 * the writing part that we have finished reading before get the query result.
+	 * However, we will not get the result until the writing part shutdown.
+	 * The writing part will wait for a notify from the reader to shutdown.
+	 * To prevent this we send a notify to writer before we receive result.
+	 */
+	ListCell   *cell;
+	foreach (cell, estate->sharedScanConsumers)
+	{
+		ShareInputScanState *state = lfirst(cell);
+		if (state->share_lk_ctxt)
```

Comment:
Is this check redundant?

Reply:
in the current version of the patch, no, because if SharedScan occurs inside one slide, it will be NULL.

Reply:
ok, I checked it looks like for non-cross slice Shared Scan driver_slice is -1 and such SharedScan is not added to the list so we can remove this condition.

Reply:
and shareinput_reader_waitready expects non-null first argument ;)